compiler/env/TypeLayout.hpp

```diff
@@ -0,0 +1,103 @@
+ * Copyright (c) 2019, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/** \brief
+ *	    The two classes track the sequence of fields so that 
+ *     the compiler can iterate through and look them up. 
+ *
+ *     TypeLayoutEntry represents a field. For each field entry its TR::DataType and 
+ *     its offset (including the object header size) are recorded. 
+ *     TypeLayout represents the layout of fields containing TypeLayoutEntry that belongs to 
+ *     a single class. It stores a `TR::vector` of TypeLayoutEntry objects. 
+ */
+#ifndef TYPE_LAYOUT_INCL
+#define TYPE_LAYOUT_INCL
+
+#include "il/DataTypes.hpp"
+#include "infra/vector.hpp"
+
+namespace TR {
+
+class TypeLayoutEntry
+   {
+   public:
+   TypeLayoutEntry(TR::DataType datatype, int32_t offset, char * fieldname) 
+      : _datatype(datatype), _offset(offset), _fieldname(fieldname){}
+   TR::DataType _datatype;
+   int32_t _offset;
+   const char * _fieldname;
+   };
+
+class TypeLayout
+   {
+   public:
+   TypeLayout(TR::Region& region) : _entries(region) {}
+   void add(TR::TypeLayoutEntry entry) { _entries.push_back(entry); }
+   void sort() 
+      { 
+      struct compOffset
+         {
+         bool operator() (const TypeLayoutEntry& entry1, const TypeLayoutEntry& entry2) const 
+            { 
+            return entry1._offset < entry2._offset; 
+            }
+         } compareOffset;
+      std::sort(_entries.begin(), _entries.end(), compareOffset);
+      }
+   size_t count() const { return _entries.size(); }
+
+   /** \brief
+    *     Finds the corresponding entry based on index given.
+    * 
+    *  \param ind
+    *     The index of the entry that wants to be returned.
+    * 
+    *  \return
+    *     The TypeLayoutEntry object stored in the vector at given index.
+    */
+   const TypeLayoutEntry& entry(size_t ind) const { return _entries.at(ind); } 
```

Comment:
Here does index mean that we just number the fields in offset order? how does one go from a field to an index I think is a logical question that I am not quite clear on...